libraries/tock-cells/src/volatile_cell.rs

```diff
@@ -1,29 +1,81 @@
 //! Implementation of a type for accessing MCU registers.
 
+use core::cell::UnsafeCell;
 use core::ptr;
 
 /// `VolatileCell` provides a wrapper around unsafe volatile pointer reads
 /// and writes. This is particularly useful for accessing microcontroller
-/// registers.
-// Source: https://github.com/hackndev/zinc/tree/master/volatile_cell
-#[derive(Copy, Clone, Default)]
-#[repr(C)]
-pub struct VolatileCell<T> {
-    value: T,
+/// registers by (unsafely) casting a pointer to the register into a `VolatileCell`.
+///
+/// ```
+/// use tock_cells::volatile_cell::VolatileCell;
+/// let myptr: *const usize = 0xdeadbeef as *const usize;
+/// let myregister: &VolatileCell<usize> = unsafe { core::mem::transmute(myptr) };
+/// ```
+// Originally modified from: https://github.com/hackndev/zinc/tree/master/volatile_cell
+#[derive(Default)]
+#[repr(transparent)]
+pub struct VolatileCell<T: ?Sized + Copy> {
+    value: UnsafeCell<T>,
 }
 
-impl<T> VolatileCell<T> {
+impl<T: Copy> Clone for VolatileCell<T> {
+    #[inline]
+    fn clone(&self) -> Self {
+        VolatileCell::new(self.get())
+    }
+}
+
+impl<T: Copy> VolatileCell<T> {
     pub const fn new(value: T) -> Self {
-        VolatileCell { value }
+        VolatileCell {
+            value: UnsafeCell::new(value),
+        }
     }
 
     #[inline]
-    pub fn get(&self) -> T {
-        unsafe { ptr::read_volatile(&self.value) }
+    /// Performs a memory write with the provided value.
+    ///
+    /// # Side-Effects
+    ///
+    /// `set` _always_ performs a memory write on the underlying location. If
+    /// this location is a memory-mapped I/O register, the side-effects of
+    /// performing the write are register-specific.
+    ///
+    /// `set` does not read the value currently inside the `VolatileCell` and,
```

Comment:
nit: With the `T: Copy` restriction, this line isn't necessary, at least in the doc string. Maybe instead have a comment inside the function that mentions that because `T: Copy`, there's no drop to consider?